### ISSUE TITLE
Compact definition of dm objects

### ISSUE THREAD

**Issue by moodymudskipper** [cynkra/dm#814]
dm objects are long and verbose to define, because of this : 
* definitions are not easy to grasp at a glance
* it's long to type
* might be barrier to entry for new users
* it's not easy to define simple prototypes

The last point is the one that brought the ideas that I show here. We use prototype objects in `dm_unnest_tbl()` and `dm_unpack_tbl()` and for consistency they might be used in other functions discussed in https://github.com/cynkra/dm/pull/223#pullrequestreview-861105775

@krlmlr I believe you mentioned that you'd like to be able to define dms without worrying about data types, this is a step in that direction.

I'd like to take inspiration from `tibble::tribble()`, this function is a bit weird but it's so handy that this is quickly forgiven. 

Take the following dm code :

```
dm(
  airlines = tibble(carrier = NA),
  airports = tibble(origin = NA, faa = NA),
  flights = tibble(tailnum = NA, carrier = NA, origin = NA),
  planes = tibble(tailnum = NA),
  weather = tibble() 
) %>% 
  dm_add_pk(table = airlines, columns = carrier) %>%
  dm_add_pk(airports, faa) %>%
  dm_add_pk(planes, tailnum) %>%
  dm_add_fk(table = flights, columns = tailnum, ref_table = planes) %>%
  dm_add_fk(flights, carrier, airlines) %>%
  dm_add_fk(flights, origin, airports)
```

It might be reproduced by something like :

```
quick_dm(
  ~airlines , ~airports,~flights           ,~planes   ,  ~weather # tables
  ,"carrier",     "faa",                   , "tailnum",           # pks 
  ,         ,          , tailnum ~ planes  ,          ,           # fks
  ,         ,          , carrier ~ airlines,          ,           # fks
  ,         ,          , origin ~ airports ,          ,           # fks
)
```

If we do have the data frames we don't use formulas :  

```
quick_dm(
  airlines  ,airports  ,flights            ,planes    ,weather # tables
  ,"carrier",     "faa",                   , "tailnum",        # pks 
  ,         ,          , tailnum ~ planes  ,          ,        # fks
  ,         ,          , carrier ~ airlines,          ,        # fks
  ,         ,          , origin ~ airports ,          ,        # fks
)
```

The `...` can contain one sided formulas, data frames, character and 2 sided formulas so it's unambiguous.

This is not as general as it could be but with a bit of imagination we might be able to deal with multi column keys and foreign keys pointing to explicit keys.

-------------------

My original take looked like this, I prefer it personally, but it uses  NSE with `->>` and `$`, which might be too weird ?

```
quick_dm(
  # weather needs to be defined explicitly since not linked to other tables
  # the other tables can be inferred from relationships
  ~weather, 
  airports$faa, # define pk
  flights$tailnum ->> planes,  # define fk and associated pk      
  flights$carrier ->> airlines$carrier, # mention pk explicitly
  flights["origin"] ->> airports["origin"] # another way to do it, we might have several cols here           
)
```

Here also if we don't use formulas we can provide actual data frames (or create them on the go as with `dm()`), then we don't fetch the columns in the rest of the code to create them.

We can use multi column keys with `[`

We can use `->` rather than `->>` for foreign keys pointing to non pk reference columns

If `->` and `->>` are taboo we also have `:=`, `~`, `==`, `>`, `?` , `/`, `&`, `|`, `:`

-------------

Using any of those or a variant, we can define 2 table prototypes very easily for `dm_unnest_tbl()` for instance using the latter`quick_dm(child$fk ->> parent$pk)` is a valid 2 tables dm with a pk and a fk.

-------------

@krlmlr @TSchiefer Would it be useful ?

**Comment by krlmlr** [comment]
Is this a duplicate of https://github.com/cynkra/dm/issues/748?

**Comment by moodymudskipper** [comment]
Right, In the spirit it is I think, at least it does overlap but :

* I personally like the idea of building a dm with a single call, somewhat similar to what `ggplot2::qplot()` does (though I know its use is discouraged).
* #748 makes `dm_add_fk()`  and `dm_add_fk()` less strict, so typos would be accepted silently, unless we add an argument, but then they're more verbose

Could both features coexist ?

**Comment by moodymudskipper** [comment]
We have 2 use cases here : 
* (1) generally building dms quicker
* (2) providing convenient `ptype` args to functions from https://github.com/cynkra/dm/issues/817 

If we cannot find a satisfactory way to do (1) and (2) together,  we might have one or more specific functions designed to work only as values of `ptype`, these might work like tidy select functions, i.e. use context by default. The scope being much narrower these would probably be easier to design.

**Comment by moodymudskipper** [comment]
I've drafted `quick_dm()` in https://github.com/cynkra/dm/tree/feature/quick_dm , hoping to play a bit with it and use it as a base for discussion.

``` r
library(dm)
#> 
#> Attaching package: 'dm'
#> The following object is masked from 'package:stats':
#> 
#>     filter
# minimal dms
dm1 <- quick_dm(child$fk ->> parent$pk)
dm_draw(dm1)
```

![](https://i.imgur.com/Delabtm.png)

``` r
dm2 <- quick_dm(~ table_without_key, NULL ->> table_with_pk$pk)
dm_draw(dm2)
```

![](https://i.imgur.com/CTElBnF.png)

``` r
# reproduce structure of some dms from the package
dm_nycflights13_quick <-
  quick_dm(
    flights$carrier ->> airlines,
    flights$tailnum ->> planes,
    flights$origin ->> airports,
    flights[c("origin", "time_hour")] ->> weather,
  )
dm_draw(dm_nycflights13_quick)
```

![](https://i.imgur.com/HB0Z23f.png)

``` r
dm_for_filter_quick <-
  quick_dm(
    tf_2$d ->> tf_1$a,
    NULL ->> tf_2$c,
    tf_2[c("e","e1")] ->> tf_3[c("f", "f1")],
    tf_4[c("j","j1")] ->> tf_3[c("f", "f1")],
    tf_5$l ->> tf_4$h,
    tf_5$m -> tf_6$n,
    NULL ->> tf_5$k,
    NULL ->> tf_6$o
  )
dm_draw(dm_for_filter_quick)
```

![](https://i.imgur.com/LnI7tC4.png)

``` r
# building prototypes for `dm_unnest_tbl`
airlines_wrapped <-
  dm_nycflights13() %>%
  dm_wrap_tbl(airlines)

airlines_wrapped %>%
  dm_unnest_tbl(airlines, flights, quick_dm(flights$carrier ->> airlines))
#> ── Metadata ────────────────────────────────────────────────────────────────────
#> Tables: `airlines`, `flights`
#> Columns: 24
#> Primary keys: 1
#> Foreign keys: 1
```

<sup>Created on 2022-02-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
